Feature/182 playbook conversion

thijshberg opened the pull request:
This adds conversion from bpmn to cacao. Bpmn does not have the variables or actions that cacao (or soarca) has, so this is effectively a starting point for real conversion.

thijshberg committed:
Added convertor command and initial work on BPMN convertor

thijshberg committed:
Added format tests and BPMN test file

thijshberg committed:
Simplest BPMN file conversion now working

thijshberg committed:
Added control gate support to BPMN conversion

thijshberg committed:
Added required fields to converted files to conform to cacao validator

thijshberg committed:
Updated to process all compatible CISA examples

These are examples of BPMN playbooks. The only unsupported elements in
there are:
 - "intermediateThrowEvent", "intermediateCatchEvent": these require
   storage/communication between playbooks. We can make a playbook step
   in cacao, but that's not the same thing (and requires context).
 - "inclusiveGateway": usually used for a choice, e.g. "what is the
   course of action". Can be converted using manual steps, but the
   examples lack text in the gateway to give enough data to do this.

The choice for both of these for now is to error and not generate
output. There should be more diagnostic output in general on what needs
to be done for a meaningful conversion.

thijshberg committed:
Updated tests to check for valid JSON output

thijshberg committed:
Added more metadata to converted file

This metadata should still be changed by the implementer, as this is not
taken from the BPMN playbook.

This also does not manage to please Cymph, which does not find this a valid
playbook, but refuses to explain why.

thijshberg committed:
Fixed linter errors

thijshberg committed:
Removed unused formats

thijshberg committed:
Added tests

thijshberg committed:
fixed lint errors

MaartendeKruijf commented on makefile:
```diff
@@ -14,11 +14,11 @@ swagger:
 
 build: swagger
 	CGO_ENABLED=0 go build -o ./build/soarca $(GOFLAGS) ./cmd/soarca/main.go
+	CGO_ENABLED=0 go build -o ./build/soarca-conversion $(GOFLAGS) ./cmd/soarca-conversion/main.go
```
Make this into another command

MaartendeKruijf commented on pkg/conversion/bpmn/bpmn_conversion.go:
```diff
@@ -0,0 +1,415 @@
+package conversion
```
Move file to pkg/conversion/bpmn

MaartendeKruijf commented on pkg/conversion/perform_conversion.go:
```diff
@@ -0,0 +1,28 @@
+	"soarca/pkg/models/cacao"
+)
+
+func PerformConversion(input_filename string, input []byte, format_string string) (*cacao.Playbook, error) {
+	var format TargetFormat
+	if format_string == "" {
+		format = guess_format(input_filename)
+	} else {
+		format = read_format(format_string)
+	}
+	if format == FormatUnknown {
+		return nil, errors.New("could not deduce input file type")
+	}
+	var converter IConverter
+	switch format {
+	case FormatBpmn:
+		converter = NewBpmnConverter()
+	}
+	return converter.Convert(input, input_filename)
+}
```
Only have interface definition in this file

MaartendeKruijf commented on pkg/conversion/format.go:
```diff
@@ -0,0 +1,24 @@
+		return FormatBpmn
+	}
+	return FormatUnknown
+}
```
Move to util

MaartendeKruijf commented on pkg/conversion/bpmn_conversion.go:
```diff
@@ -0,0 +1,415 @@
+	annotations []BpmnAnnotation
+}
+
+func (p *BpmnProcess) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
```
p → process d → decode

MaartendeKruijf commented on pkg/conversion/bpmn_conversion.go:
```diff
@@ -0,0 +1,415 @@
+		cacao.AgentTarget{
+			ID:   fmt.Sprintf("individual--%s", uuid.New()),
+			Type: "individual",
+			Name: "CHANGE THIS"})
```
Just leave it empty

MaartendeKruijf commented on pkg/conversion/bpmn/bpmn_conversion.go:
```diff
@@ -0,0 +1,415 @@
+	Id       string `xml:"id,attr"`
+	Incoming string `xml:"bpmn:incoming"`
+}
+type BpmnTask struct {
+	Kind string
+	Id   string `xml:"id,attr"`
+	Name string `xml:"name,attr"`
+}
+
+type BpmnFlow struct {
+	Id            string `xml:"id,attr"`
+	SourceRef     string `xml:"sourceRef,attr"`
+	TargetRef     string `xml:"targetRef,attr"`
+	Name          string `xml:"name,attr"`
+	IsAssociation bool
+}
```
Can we collapse the types into a few, if not then that is fine

MaartendeKruijf commented on cmd/soarca-conversion/main.go:
```diff
@@ -0,0 +1,76 @@
+	flag.StringVar(&source_filename, "source", "", "The source file to be converted")
+	flag.StringVar(&target_filename, "target", "", "The name of the converted filename")
+	flag.StringVar(&format, "format", "", "The format of the source file")
+	flag.Parse()
+	if source_filename == "" {
+		log.Error("No source file given: -source=SOURCE_FILE is required")
+		return
+	}
```
Maybe also add a help as this is a command line tool

MaartendeKruijf commented on pkg/conversion/bpmn_conversion.go:
```diff
@@ -0,0 +1,415 @@
+	if len(definitions.Processes) == 0 {
+		return nil, errors.New("BPMN file does not have any processes")
+	}
+	playbook := cacao.NewPlaybook()
```
Maybe we should make a default playbook init command in the cacao implementation

MaartendeKruijf commented on pkg/conversion/bpmn/bpmn_conversion.go:
```diff
@@ -0,0 +1,415 @@
+
+func (p *BpmnProcess) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
+	start_name := start.Name
+	for {
```
can't we directly unmarshal this into a structure like we do with the JSON playbooks.

thijshberg replied:
This is commented in the code (same goes for the large-ish collection of types we end up with): the BPMN process is a non-homogenous list of tasks, control gates, annotations, and connections between them called flows. Putting all of those in one shared type would lead to code with a lot of assumptions that are complicated to deal with and can cause issues at runtime. The eyesore of having some types to deal with and a large switch for parsing are probably easier to deal with.

MaartendeKruijf commented on pkg/conversion/bpmn_conversion.go:
```diff
@@ -0,0 +1,415 @@
+		Type:      cacao.StepTypeIfCondition,
+		Condition: gateway.Name,
+	}
+	name := fmt.Sprintf("if-condition--%s", uuid.New())
```
stepId

MaartendeKruijf commented on pkg/conversion/bpmn_conversion.go:
```diff
@@ -0,0 +1,415 @@
+	return nil
+}
+
+func (end_event BpmnEndEvent) implement(playbook *cacao.Playbook, converter *BpmnConverter) error {
```
We can generalise the steps right? The steps are all of type cacao.Step, so I would create builder functions for each step type. And just add them outside of the function.

MaartendeKruijf commented on pkg/conversion/bpmn_conversion.go:
```diff
@@ -0,0 +1,415 @@
+	playbook.Type = "playbook"
+	playbook.ID = fmt.Sprintf("playbook--%s", uuid.New())
+	playbook.CreatedBy = fmt.Sprintf("identity--%s", uuid.New())
+	playbook.Name = clean_filename(filename)
```
Shouldn't we use the process name?

MaartendeKruijf commented on test/conversion/control_gates.bpmn:
```diff
@@ -0,0 +1,153 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<bpmn:definitions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" id="Definitions_08y6v5x" targetNamespace="http://bpmn.io/schema/bpmn" exporter="bpmn-js (https://demo.bpmn.io)" ex ...
```
Maybe also use a playbook from https://github.com/cisagov/shareable-soar-workflows

MaartendeKruijf commented on test/conversion/conversion_test.go:
```diff
@@ -0,0 +1,87 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func loadPlaybook(t *testing.T, filename string) *cacao.Playbook {
+	input, err := os.ReadFile(filename)
+	assert.Nil(t, err)
+	playbook, err := conversion.NewBpmnConverter().Convert(input, filename)
+	assert.Nil(t, err)
+	return playbook
+}
+
+func nextSteps(t *testing.T, step cacao.Step, playbook *cacao.Playbook) []cacao.Step {
+	var steps []cacao.Step
+	for _, step_name := range step.NextSteps {
+		steps = append(steps, findStep(t, step_name, playbook))
+	}
+	return steps
+}
+func findStep(t *testing.T, step_name string, playbook *cacao.Playbook) cacao.Step {
+	step, ok := playbook.Workflow[step_name]
+	assert.True(t, ok, "Could not find %s", step_name)
+	return step
+}
```
These functions complicate the testing logic and should be ideally tested aswel. I would make the functions with direct assertions.

MaartendeKruijf commented on test/conversion/conversion_test.go:
```diff
@@ -0,0 +1,87 @@
+	assert.True(t, ok, "Could not find %s", step_name)
+	return step
+}
+func findStepByName[S ~[]cacao.Step](t *testing.T, step_name string, steps S) *cacao.Step {
+	for _, step := range steps {
+		if step.Name == step_name {
+			return &step
+		}
+	}
```
step names are used as variables but these apear to be ids so you can just do a lookup in the workflows map

thijshberg replied:
This function looks up by the name field, not Id's (that's the function above it). This is very convenient to find the converted step knowing only the original name.

MaartendeKruijf reviewed (changes requested):
Nice to see this coming along, some changes are needed but good progress

thijshberg committed:
Update makefile rules

This puts both binaries into a simple rule. This does complicate things
if we add other targets in build/, but for now this works fine.

thijshberg committed:
Small changes based on code review

This includes some points where code was put in more correct locations, and some extra testing.

thijshberg committed:
Update makefile to have one go build rule

This works for both build/soarca and build/soarca-conversion, and
properly depends on all go files to detect rebuilds. The default build
target always triggers this, as swag will touch api/docs.go; filtering
this out broke the dependency for reasons beyond comprehension, but
since builds are quite quick this is okay for now.
